library/pem.c

```diff
@@ -491,7 +491,7 @@ int mbedtls_pem_write_buffer(const char *header, const char *footer,
     size_t len = 0, use_len, add_len = 0;
 
     mbedtls_base64_encode(NULL, 0, &use_len, der_data, der_len);
-    add_len = strlen(header) + strlen(footer) + (use_len / 64) + 1;
+    add_len = strlen(header) + strlen(footer) + (((use_len > 2) ? (use_len - 2) : 0) / 64) + 1;
```

Comment:
I'm sorry, but I can't make sense of this `- 2`. I believe the code is correct, as I've read the test code and is seems correct and thorough and it's passing so the code must be behaving correctly, but I need to understand the code as well before I can approve.

Reply:
I've investigated a bit and I found the root cause is actually one layer deeper. In the test function's loop, values taken by `use_len` here are 0, 5, 9, 13, etc. which is suspect. Values after the 2nd call to `mbedtls_bas64_encode()` (a few lines below) are 0, 4, 8, 12 which is much more logical considering how base64 encoding works. The documentation of `mbedtls_base64_encode()` clearly suggests that it's supposed to set `olen` to the same value in both cases, so that function is buggy. Looking at the code, the function adds a nul byte at the end but doesn't always include it in the length count.

Considering this PR is already an interrupt from #6838 (which hits the sensitive value of 48 for `der_len`), I don't think we want to dig too deep into questions about what `base64_encode()` should actually do and how to fix in a way the preserves compatibility right now, so I'm approving but creating a follow-up issue.

Reply:
See https://github.com/Mbed-TLS/mbedtls/issues/7550